File: apps/dsa-web/src/api/screener.ts
import apiClient from './index';

export type ScreenerStrategyId =
  | 'quality_value'
  | 'multi_factor'
  | 'trend_follow'
  | 'pullback'
  | 'breakout';

export type ScreenerStrategyInfo = {
  id: ScreenerStrategyId;
  name: string;
  description: string;
  cadence: string;
  dataScope: string[];
  iwencaiFit: string;
};

export type ScreenerStrategyLibraryItem = {
  id: string;
  name: string;
  description: string;
  query: string;
  backtestStatus: string;
  lastRunResult?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type ScreenerStrategyLibraryUpsert = {
  name: string;
  description: string;
  query: string;
  backtestStatus?: string;
  lastRunResult?: string | null;
};

export type ScreenerCandidate = {
  code: string;
  name?: string | null;
  score: number;
  matchedStrategies: string[];
  reasons: string[];
  risks: string[];
  metrics: Record<string, number | null | undefined>;
  iwencaiFields: Record<string, string>;
  latestDate?: string | null;
  dataSource: string;
};

export type ScreenerRunRequest = {
  strategyIds: ScreenerStrategyId[];
  stockCodes?: string[];
  iwencaiQuery?: string;
  iwencaiPage: number;
  limit: number;
  includeFundamentals: boolean;
  useIwencai: boolean;
  strategyLibraryId?: string;
};

export type ScreenerRunResponse = {
  strategies: ScreenerStrategyInfo[];
  candidates: ScreenerCandidate[];
  totalInput: number;
  evaluated: number;
  skipped: number;
  dataMode: string;
  executionMode: string;
  localExecutable: boolean;
  supportedTerms: string[];
  unsupportedTerms: string[];
  importRequired: boolean;
  iwencaiStatus: string;
  iwencaiQuery?: string | null;
  iwencaiCodeCount?: number | null;
  iwencaiReturnedCount?: number | null;
  iwencaiHasMore: boolean;
  iwencaiChunksInfo: Record<string, unknown>;
  notes: string[];
};

type ApiStrategyInfo = {
  id: ScreenerStrategyId;
  name: string;
  description: string;
  cadence: string;
  data_scope?: string[];
  iwencai_fit?: string;
};

type ApiCandidate = {
  code: string;
  name?: string | null;
  score: number;
  matched_strategies?: string[];
  reasons?: string[];
  risks?: string[];
  metrics?: Record<string, number | null | undefined>;
  iwencai_fields?: Record<string, string>;
  latest_date?: string | null;
  data_source?: string;
};

type ApiRunResponse = {
  strategies?: ApiStrategyInfo[];
  candidates?: ApiCandidate[];
  total_input: number;
  evaluated: number;
  skipped: number;
  data_mode: string;
  execution_mode?: string;
  local_executable?: boolean;
  supported_terms?: string[];
  unsupported_terms?: string[];
  import_required?: boolean;
  iwencai_status: string;
  iwencai_query?: string | null;
  iwencai_code_count?: number | null;
  iwencai_returned_count?: number | null;
  iwencai_has_more?: boolean;
  iwencai_chunks_info?: Record<string, unknown>;
  notes?: string[];
};

type ApiStrategyLibraryItem = {
  id: string;
  name: string;
  description: string;
  query: string;
  backtest_status: string;
  last_run_result?: string | null;
  created_at: string;
  updated_at: string;
};

function mapStrategy(item: ApiStrategyInfo): ScreenerStrategyInfo {
  return {
    id: item.id,
    name: item.name,
    description: item.description,
    cadence: item.cadence,
    dataScope: item.data_scope ?? [],
    iwencaiFit: item.iwencai_fit ?? '',
  };
}

function mapLibraryItem(item: ApiStrategyLibraryItem): ScreenerStrategyLibraryItem {
  return {
    id: item.id,
    name: item.name,
    description: item.description,
    query: item.query,
    backtestStatus: item.backtest_status,
    lastRunResult: item.last_run_result,
    createdAt: item.created_at,
    updatedAt: item.updated_at,
  };
}

function mapCandidate(item: ApiCandidate): ScreenerCandidate {
  return {
    code: item.code,
    name: item.name,
    score: item.score,
    matchedStrategies: item.matched_strategies ?? [],
    reasons: item.reasons ?? [],
    risks: item.risks ?? [],
    metrics: item.metrics ?? {},
    iwencaiFields: item.iwencai_fields ?? {},
    latestDate: item.latest_date,
    dataSource: item.data_source ?? '',
  };
}

function mapRunResponse(data: ApiRunResponse): ScreenerRunResponse {
  return {
    strategies: (data.strategies ?? []).map(mapStrategy),
    candidates: (data.candidates ?? []).map(mapCandidate),
    totalInput: data.total_input,
    evaluated: data.evaluated,
    skipped: data.skipped,
    dataMode: data.data_mode,
    executionMode: data.execution_mode ?? 'local_query',
    localExecutable: data.local_executable ?? true,
    supportedTerms: data.supported_terms ?? [],
    unsupportedTerms: data.unsupported_terms ?? [],
    importRequired: data.import_required ?? false,
    iwencaiStatus: data.iwencai_status,
    iwencaiQuery: data.iwencai_query,
    iwencaiCodeCount: data.iwencai_code_count,
    iwencaiReturnedCount: data.iwencai_returned_count,
    iwencaiHasMore: data.iwencai_has_more ?? false,
    iwencaiChunksInfo: data.iwencai_chunks_info ?? {},
    notes: data.notes ?? [],
  };
}

export const screenerApi = {
  async getStrategies(): Promise<ScreenerStrategyInfo[]> {
    const response = await apiClient.get('/api/v1/screener/strategies');
    return (response.data as ApiStrategyInfo[]).map(mapStrategy);
  },

  async run(request: ScreenerRunRequest): Promise<ScreenerRunResponse> {
    const response = await apiClient.post('/api/v1/screener/run', {
      strategy_ids: request.strategyIds,
      stock_codes: request.stockCodes,
      iwencai_query: request.iwencaiQuery,
      iwencai_page: request.iwencaiPage,
      limit: request.limit,
      include_fundamentals: request.includeFundamentals,
      use_iwencai: request.useIwencai,
      strategy_library_id: request.strategyLibraryId,
    });
    return mapRunResponse(response.data as ApiRunResponse);
  },

  async getLibrary(): Promise<ScreenerStrategyLibraryItem[]> {
    const response = await apiClient.get('/api/v1/screener/library');
    return (response.data as ApiStrategyLibraryItem[]).map(mapLibraryItem);
  },

  async createLibraryItem(request: ScreenerStrategyLibraryUpsert): Promise<ScreenerStrategyLibraryItem> {
    const response = await apiClient.post('/api/v1/screener/library', {
      name: request.name,
      description: request.description,
      query: request.query,
      backtest_status: request.backtestStatus,
      last_run_result: request.lastRunResult,
    });
    return mapLibraryItem(response.data as ApiStrategyLibraryItem);
  },

  async updateLibraryItem(id: string, request: ScreenerStrategyLibraryUpsert): Promise<ScreenerStrategyLibraryItem> {
    const response = await apiClient.put(`/api/v1/screener/library/${encodeURIComponent(id)}`, {
      name: request.name,
      description: request.description,
      query: request.query,
      backtest_status: request.backtestStatus,
      last_run_result: request.lastRunResult,
    });
    return mapLibraryItem(response.data as ApiStrategyLibraryItem);
  },

  async importIwencaiExcel(params: {
    file: File;
    strategyQuery?: string;
    strategyLibraryId?: string;
    limit?: number;
  }): Promise<ScreenerRunResponse> {
    const form = new FormData();
    form.append('file', params.file);
    form.append('strategy_query', params.strategyQuery ?? '');
    form.append('strategy_library_id', params.strategyLibraryId ?? '');
    form.append('limit', String(params.limit ?? 100));
    const response = await apiClient.post('/api/v1/screener/imports/iwencai-excel', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return mapRunResponse(response.data as ApiRunResponse);
  },
};

File: api/v1/schemas/screener.py
# -*- coding: utf-8 -*-
"""Stock screener API schemas."""

from __future__ import annotations

from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field


ScreenerStrategyId = Literal[
    "quality_value",
    "multi_factor",
    "trend_follow",
    "pullback",
    "breakout",
]


class ScreenerRunRequest(BaseModel):
    strategy_ids: List[ScreenerStrategyId] = Field(
        default_factory=lambda: ["quality_value", "multi_factor", "trend_follow"],
        description="要执行的选股策略",
    )
    stock_codes: Optional[List[str]] = Field(
        default=None,
        description="候选股票池；为空时使用系统自选股列表",
    )
    iwencai_query: Optional[str] = Field(
        default=None,
        description="问财自然语言全市场选股条件；use_iwencai=true 时生效",
    )
    iwencai_page: int = Field(1, ge=1, le=100, description="问财分页页码")
    limit: int = Field(30, ge=1, le=100, description="返回结果上限")
    include_fundamentals: bool = Field(
        False,
        description="是否补充基本面数据；会比纯技术筛选更慢",
    )
    use_iwencai: bool = Field(
        False,
        description="兼容旧字段；当前主流程不再在线调用问财选股",
    )
    strategy_library_id: Optional[str] = Field(
        default=None,
        description="策略库条目 ID；传入后会更新该策略的最近运行结果",
    )


class ScreenerStrategyInfo(BaseModel):
    id: str
    name: str
    description: str
    cadence: str
    data_scope: List[str] = Field(default_factory=list)
    iwencai_fit: str


class ScreenerCandidate(BaseModel):
    code: str
    name: Optional[str] = None
    score: float
    matched_strategies: List[str] = Field(default_factory=list)
    reasons: List[str] = Field(default_factory=list)
    risks: List[str] = Field(default_factory=list)
    metrics: Dict[str, Optional[float]] = Field(default_factory=dict)
    iwencai_fields: Dict[str, str] = Field(default_factory=dict)
    latest_date: Optional[str] = None
    data_source: str = ""


class ScreenerRunResponse(BaseModel):
    strategies: List[ScreenerStrategyInfo] = Field(default_factory=list)
    candidates: List[ScreenerCandidate] = Field(default_factory=list)
    total_input: int
    evaluated: int
    skipped: int
    data_mode: str
    execution_mode: str = "local_query"
    local_executable: bool = True
    supported_terms: List[str] = Field(default_factory=list)
    unsupported_terms: List[str] = Field(default_factory=list)
    import_required: bool = False
    iwencai_status: str
    iwencai_query: Optional[str] = None
    iwencai_code_count: Optional[int] = None
    iwencai_returned_count: Optional[int] = None
    iwencai_has_more: bool = False
    iwencai_chunks_info: Dict[str, object] = Field(default_factory=dict)
    notes: List[str] = Field(default_factory=list)


class ScreenerStrategyLibraryItem(BaseModel):
    id: str
    name: str
    description: str
    query: str
    backtest_status: str = "未回测"
    last_run_result: Optional[str] = None
    created_at: str
    updated_at: str


class ScreenerStrategyLibraryUpsertRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=80)
    description: str = Field(..., min_length=1, max_length=400)
    query: str = Field(..., min_length=1, max_length=1000)
    backtest_status: Optional[str] = Field(default=None, max_length=80)
    last_run_result: Optional[str] = Field(default=None, max_length=400)
